athena-redis/src/main/java/com/amazonaws/athena/connectors/redis/lettuce/RedisConnectionFactory.java

```diff
@@ -0,0 +1,115 @@
+/*-
+ * #%L
+ * athena-redis
+ * %%
+ * Copyright (C) 2019 - 2021 Amazon Web Services
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.amazonaws.athena.connectors.redis.lettuce;
+
+import io.lettuce.core.ReadFrom;
+import io.lettuce.core.RedisClient;
+import io.lettuce.core.RedisURI;
+import io.lettuce.core.cluster.RedisClusterClient;
+import io.lettuce.core.cluster.api.StatefulRedisClusterConnection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+
+public class RedisConnectionFactory
+{
+  private static final Logger logger = LoggerFactory.getLogger(RedisConnectionFactory.class);
+
+  private static final int CONNECTION_TIMEOUT_MS = 2_000;
+  private final Map<String, RedisConnectionWrapper<String, String>> clientCache = new HashMap<>();
+
+  public synchronized RedisConnectionWrapper<String, String> getOrCreateConn(String conStr, boolean sslEnabled,
+                                                                             boolean isCluster)
+  {
+    String conKey = conStr + sslEnabled + isCluster;
+    RedisConnectionWrapper<String, String> connection = clientCache.get(conKey);
+    if (connection == null) {
+      String[] endpointParts = conStr.split(":");
+      if (endpointParts.length == 2) {
+        if (isCluster) {
+          connection = getOrCreateClusterCon(endpointParts[0], Integer.parseInt(endpointParts[1]), null, sslEnabled);
+        }
+        else {
+          connection = getOrCreateStandaloneCon(endpointParts[0], Integer.parseInt(endpointParts[1]), null, sslEnabled);
+        }
+      }
+      else if (endpointParts.length == 3) {
+        if (isCluster) {
+          connection = getOrCreateClusterCon(endpointParts[0], Integer.parseInt(endpointParts[1]), endpointParts[2],
+                                             sslEnabled);
+        }
+        else {
+          connection = getOrCreateStandaloneCon(endpointParts[0], Integer.parseInt(endpointParts[1]), endpointParts[2],
+                                                sslEnabled);
+        }
+      }
+      else {
+        throw new IllegalArgumentException("Redis endpoint format error.");
+      }
+      clientCache.put(conKey, connection);
+    }
+    return connection;
```

Comment:
do we need to 'test' this connection to ensure its still usable?

Reply:
good question, like something similar to the HBaseConnection? I'll take a look

Reply:
Just checked, Lettuce has a [ConnectionWatchdog](https://lettuce.io/core/release/api/io/lettuce/core/protocol/ConnectionWatchdog.html) that monitors each connection and automatically reconnects if the connection is lost.